### PR TITLE
Move some crossmodel facades to controller

### PR DESCRIPTION
**Pull request by wallyworld**
The main purpose here is to move some of the cross model facades to controller facades.
There's a bunch of followup work to be done to move from what was implemented originally to the latest spec. Part of that is done my necessity here. eg remove AllowedUsers from some structs; also start removing code to show offered applications as these will be removed. These changes are only visible with cross-model-relations feature flag.

QA: bootstrap with feature flag, deploy and offer a service, list and show endpoints

### CODE REVIEW

File: apiserver/crossmodel/base_test.go [hunk 1]
```diff
@@ -54,32 +53,26 @@ func (s *baseCrossmodelSuite) SetUpTest(c *gc.C) {
...
+	s.applicationDirectory = &mockApplicationDiretory{}
 
 	var err error
-	s.api, err = crossmodel.CreateAPI(s.serviceBackend, crossmodel.GetStateAccess(s.State), s.resources, s.authorizer)
+	s.api, err = crossmodel.CreateAPI(s.applicationDirectory, crossmodel.GetStateAccess(s.State), s.authorizer)
 	c.Assert(err, jc.ErrorIsNil)
 }
 
-type mockServiceBackend struct {
+type mockApplicationDiretory struct {
```
Comment (mjs) [inline apiserver/crossmodel/base_test.go]: Directory

File: apiserver/crossmodel/crossmodel.go [hunk 2]
```diff
@@ -60,44 +61,63 @@ func NewAPI(
...
+	errResults, err := api.applicationDirectory.AddOffers(offers)
+	if err != nil {
+		return params.ErrorResults{}, errors.Trace(err)
+	}
+
+	// Merge the results with any errors encountered earlier.
+	var offset int
+	for i, errResult := range errResults.Results {
+		if badOfferIndexes.Contains(i) {
+			offset += 1
```
Comment (mjs) [inline apiserver/crossmodel/crossmodel.go]: `offset++` would do

File: apiserver/crossmodel/crossmodel.go [hunk 3]
```diff
@@ -60,44 +61,63 @@ func NewAPI(
...
-		}); err != nil {
-			result[i].Error = common.ServerError(err)
+		})
+	}
+	errResults, err := api.applicationDirectory.AddOffers(offers)
+	if err != nil {
+		return params.ErrorResults{}, errors.Trace(err)
+	}
+
+	// Merge the results with any errors encountered earlier.
```
Comment (mjs) [inline apiserver/crossmodel/crossmodel.go]: The merge process is a bit hard to follow. What about establishing a map of index-in-offers.Offers to index-in-result in the first loop? Then `badOfferIndexes` wouldn't be required and the results from `AddOffers` could be slotted straight into `errResults.Results`.

File: apiserver/crossmodel/crossmodel_test.go [hunk 4]
```diff
@@ -27,133 +26,204 @@ func (s *crossmodelSuite) SetUpTest(c *gc.C) {
...
+		return params.ErrorResults{Results: []params.ErrorResult{{Error: &params.Error{Message: msg}}}}, nil
 	}
 
 	errs, err := s.api.Offer(all)
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(errs.Results, gc.HasLen, len(all.Offers))
 	c.Assert(errs.Results[0].Error, gc.ErrorMatches, fmt.Sprintf(".*%v.*", msg))
-	s.serviceBackend.CheckCallNames(c, addOfferBackendCall)
+	s.applicationDirectory.CheckCallNames(c, addOffersBackendCall)
 }
```
Comment (mjs) [inline apiserver/crossmodel/crossmodel_test.go]: Is there a test which covers some offers failing due to an offer params creation error?

File: cmd/juju/crossmodel/offer.go [hunk 5]
```diff
@@ -140,35 +125,35 @@ func (c *offerCommand) Run(_ *cmd.Context) error {
...
-		return errors.NotValidf(`application name %q`, ApplicationName)
+	if !names.IsValidApplication(c.Application) {
+		return errors.NotValidf(`application name %q`, c.Application)
 	}
-	c.Application = ApplicationName
 
-	endpoints := strings.SplitN(parts[1], ",", -1)
-	if len(endpoints) < 1 || endpoints[0] == "" {
-		return errors.Errorf(`specify endpoints for %v" `, ApplicationName)
+	c.Endpoints = strings.SplitN(endpoints, ",", -1)
```
Comment (mjs) [inline cmd/juju/crossmodel/offer.go]: `strings.Split` is the same as `SplitN` with -1

**Comment by wallyworld** [comment]
$$merge$$
